agent/agenttest/agent.go

```diff
@@ -0,0 +1,97 @@
+package agenttest
+
+import (
+	"context"
+	"net/url"
+	"sync"
+	"testing"
+
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/assert"
+
+	"cdr.dev/slog"
+	"cdr.dev/slog/sloggers/slogtest"
+	"github.com/coder/coder/v2/agent"
+	"github.com/coder/coder/v2/coderd/coderdtest"
+	"github.com/coder/coder/v2/codersdk"
+	"github.com/coder/coder/v2/codersdk/agentsdk"
+)
+
+// Agent is a small wrapper around an agent for use in tests.
+type Agent struct {
+	agent       agent.Agent
+	agentClient *agentsdk.Client
+	resources   []codersdk.WorkspaceResource
+	waiter      func(*codersdk.Client, uuid.UUID, ...string) []codersdk.WorkspaceResource
+	waitOnce    sync.Once
+}
+
+// Wait waits for the agent to connect to the workspace and returns the
+// resources for the connected workspace.
+// Calls coderdtest.AwaitWorkspaceAgents under the hood.
+// Multiple calls to Wait() are idempotent.
+func (a *Agent) Wait(client *codersdk.Client, workspaceID uuid.UUID, agentNames ...string) []codersdk.WorkspaceResource {
+	a.waitOnce.Do(func() {
+		a.resources = a.waiter(client, workspaceID, agentNames...)
+	})
+	return a.resources
+}
+
+// Client returns the agent client.
+func (a *Agent) Client() *agentsdk.Client {
+	return a.agentClient
+}
+
+// Agent returns the agent itself.
+func (a *Agent) Agent() agent.Agent {
```

Comment:
I think `agenttest.Agent.Agent()` is too many agents for me!

Thoughts on `New` returning the `agent.Agent` and `agentsdk.Client` types? And we just have global helper funcs, similar to `coderdtest`?

Reply:
I honestly actually prefer returning a wrapper type rather than what we did in coderdtest. The coderdtest package is a pain to maintain because if you want to return anything extra you have to update thousands of callers to have the new return value count. We also have 4 or so functions that return an API in coderdtest when it could just be 1 or 2 if the API was designed better.

I do agree though that the name Agent is too much

Reply:
Looks like this is only ever used to close the agent.  Instead this could have a `Close()` method that calls the actual agent `Close()`.

If we need to do other stuff to the agent, it's trivial to add to this package later.

Reply:
> I think agenttest.Agent.Agent() is too many agents for me!

+1 from me

Reply:
Just returning the agent now. If you need to set a custom client you can pass it in Options.